File: old/src/houghCircles.cpp
#include <include/houghCircles.h>

// gets hough space circles for r radii
void GetHoughSpaceCircles(const cv::Mat 	   &input, 			// input image
						  int  				   r_size,			// number of radii
				    	  int  				   min_r, 			// min radius
						  int  				   max_r, 			// max radius
						  int  				   r_step,			// radius step size
						  int  				   t_step,			// theta step size
				     	  std::vector<cv::Mat> &output) {		// hough spaces for r radii

	// initalise hough_circles output
	cv::Mat hough_space;
	hough_space.create(input.size(), cv::DataType<double>::type);

	// set current radius to min radius
	int radius = min_r;

	// for each radii
	for (int r=0; r<r_size; r++) {

		// reset hough space for current radius
		for (int y=0; y<input.rows; y++) {	
			for(int x=0; x<input.cols; x++) {
				hough_space.at<double>(y, x) = 0.0;
			}
		}

		// for each pixel in image
		for (int y=0; y<input.rows; y++) {	
			for(int x=0; x<input.cols; x++) {

				// get gradient magnitude at (i,j)
				double pixel_val = input.at<double>(y, x);
				
				// if gradient magnitude is not 0
				if (pixel_val > 0) {
				
					// for circle around point
					for (int theta=0; theta<360; theta+=t_step){
				
						// get circle coordinates at and angle theta
						double x0 = x - (radius * cos(theta*CV_PI/180));
						double y0 = y - (radius * sin(theta*CV_PI/180));
				
						// increment value in hough space (unless out of image)
						if (x0>=0 && y0>=0 && x0<input.cols && y0<input.rows) {
							hough_space.at<double>(y0, x0)++;
						}

					}

				}

			}
		}

		// add hough space for r and increase radius
		output.push_back(hough_space.clone());		
		radius += r_step;

	}

	std::cout << "\nHough transform circles complete!" << std::endl;
}

// sums each hough space into one summed hough space
void SumHoughSpaceCircles(std::vector<cv::Mat> &input,			// hough spaces
						  cv::Mat 			   &output) {		// summed hough space

    // set summed space as 
    output.create(input[0].size(), cv::DataType<double>::type);

    // set hough space sum to 0
    for (int y=0; y<output.rows; y++) {	
        for(int x=0; x<output.cols; x++) {
            output.at<double>(y, x) = 0.0;
        }
    }

    // for each hough space add value to summed hough space
    for (auto space : input) {
        for (int y=0; y<output.rows; y++) {
            for (int x=0; x<output.cols; x++) {
                output.at<double>(y, x) += space.at<double>(y, x);
            }
        }
    }

}

// threshold each hough space
void ThresholdHoughSpace(std::vector<cv::Mat> &input, 			// hough spaces
					 	 int 				  threshold,		// threshold value
						 std::vector<cv::Mat> &output) {		// thresholded hough spaces


	// for each hough space / radii
	int input_size = input.size();
	for (int d=0; d<input_size; d++) {

		// threshold each hough space
		cv::Mat thresholded_input;
		Threshold(input[d], threshold, thresholded_input);
		output.push_back(thresholded_input);

	}	

}

File: darts/transformation/houghcircles.py
import math
import numpy as np
from tqdm import tqdm
from darts.manipulation.utils import threshold

class HoughCircles():
    def __init__(self, min_r=40, max_r=100, r_step=5):
        self.min_r = min_r
        self.max_r = max_r
        self.r_step = r_step
        self.r_size = int((max_r-min_r) / r_step + 1)
        self.hough_space = []
        self.t_hough_space = []
        self.hough_space_sum = []
        self.t_hough_space_sum = []

    # TRY AND INCREMENT ALL HOUGHSPACE AT ONCE (np.add.at for 3D array)
    def transform(self, mag, dir, dt_size=2):
        """
        Apply Hough circles transformation to frame for a range of radii
        """

        # create r houghspaces for each radii r
        rows, cols = mag.shape
        self.hough_space = np.zeros((self.r_size, rows, cols))

        # precompute sin(theta), cos(theta) for all theta
        max_theta = 180
        thetas = np.arange(-max_theta, max_theta) * math.pi / 180
        cossin = np.column_stack((np.sin(thetas), np.cos(thetas)))

        # get positions of non-zero magnitude pixels and set initial radius
        points = np.column_stack(np.nonzero(mag))
        radius = self.min_r
        
        # progress bar
        pbar = tqdm(total=((len(points)) * (self.r_size)))
        for r in range(self.r_size):
            for (y, x) in points:
                dir_theta = np.rad2deg(dir[y][x]).astype('int')
                dt = (dir_theta - dt_size + max_theta, dir_theta + dt_size + max_theta)

                # compute circle points (y0, x0): y0=r-ysin(theta), x0=x-rcos(theta)
                c_points = np.column_stack((y - cossin[dt[0]:dt[1], 0] * radius, 
                                            x - cossin[dt[0]:dt[1], 1] * radius)).astype('int')
                # remove circle points that do not fit in frame
                c_points = c_points[np.where(~(c_points < 0).any(1) 
                                            & (c_points[:, 0] < rows) 
                                            & (c_points[:, 1] < cols))]
                # increment circle points in frame
                np.add.at(self.hough_space[r], (c_points[:, 0], c_points[:, 1]), 1)
                # update progress bar
                pbar.update(1)
            # increment radius
            radius += self.r_step

    def sum(self):
        """
        Sum hough space circles into one summed Hough space.
        """
        self.hough_space_sum = np.sum(self.t_hough_space, axis=0)

    def threshold(self, threshold_val):
        """
        Threshold each hough space, as well as the sum of Hough spaces.
        """
        self.t_hough_space = np.zeros(self.hough_space.shape)
        for r, space in enumerate(self.hough_space):
            self.t_hough_space[r] = threshold(space, threshold_val)

        # self.t_hough_space_sum = threshold(self.hough_space_sum, sum_threshold_val)